### PR TITLE
ICU-20941 Support Mixed Units in NumberFormatter when using usage()

### PR DESCRIPTION
**Pull request by hugovdm**
##### Checklist

- [X] Issue filed: https://unicode-org.atlassian.net/browse/ICU-20941
- [X] Updated PR title and link in previous line to include Issue number
- [X] Issue accepted
- [X] Tests included
- [ ] Documentation is changed or added

### CODE REVIEW

File: icu4c/source/i18n/number_microprops.h [hunk 1]
```diff
@@ -41,22 +41,55 @@ struct MicroProps : public MicroPropsGenerator {
...
+    // play.
     MeasureUnit outputUnit;
 
+    // Constant for the size of the mixedMeasures array.
+    static const int MIXED_MEASURES_LENGTH = 8;
+    // In the case of mixed units, this is the set of integer-only units
+    // *preceding* the final unit. (Simple array because it must be
+    // copy-constructable - thus can't simply use MaybeStackVector<Measure> and
+    // LocalArray<Measure>.)
+    // * TODO(units): consider dropping full "Measure" here. Use int maybe? One
```
Comment (sffc) [inline icu4c/source/i18n/number_microprops.h:86]: +1 on an array of int64s.  Also consider a helper struct to encapsulate these fields.

File: icu4c/source/i18n/number_skeletons.h [hunk 2]
```diff
@@ -246,6 +246,10 @@ void generateMeasureUnitOption(const MeasureUnit& measureUnit, UnicodeString& sb
 
 void parseMeasurePerUnitOption(const StringSegment& segment, MacroProps& macros, UErrorCode& status);
 
+/**
+ * Parses unit identifiers like "meters-per-second" and "foot-and-inch", as
```
Comment (sffc) [inline icu4c/source/i18n/number_skeletons.h:250]: meter-per-second

File: icu4c/source/i18n/number_usageprefs.h [hunk 3]
```diff
@@ -34,6 +36,12 @@ class U_I18N_API UsagePrefsHandler : public MicroPropsGenerator, public UMemory
     /**
      * Obtains the appropriate output value, MeasurementUnit and
      * rounding/precision behaviour from the UnitsRouter.
+     *
+     * The output unit is passed on to the LongNameHandler via
+     * micros.outputUnit.
+     *
+     * FIXME: For Mixed units, if we had LongNameHandler logic here, we would
```
Comment (sffc) [inline icu4c/source/i18n/number_usageprefs.h:43]: I think for mixed units without usage, you should make another handler similar to UsageHandler that populates micros.mixedMeasures.

File: icu4c/source/i18n/number_usageprefs.h [hunk 4]
```diff
@@ -52,6 +60,7 @@ class U_I18N_API UsagePrefsHandler : public MicroPropsGenerator, public UMemory
   private:
     UnitsRouter fUnitsRouter;
     const MicroPropsGenerator *fParent;
+    // LocalPointer<ListFormatter> listFormatter;
```
Comment (sffc) [inline icu4c/source/i18n/number_usageprefs.h:63]: Delete this

File: icu4c/source/i18n/unicode/measfmt.h [hunk 5]
```diff
@@ -193,6 +193,8 @@ class U_I18N_API MeasureFormat : public Format {
      * @return appendTo reference
      *
      * @stable ICU 53
+     *
+     * TODO(units): add support for mixed units.
```
Comment (sffc) [inline icu4c/source/i18n/unicode/measfmt.h:197]: Add tests for the behavior, and if the behavior is reasonable, I don't see a reason to touch this function

File: icu4c/source/test/intltest/numbertest_api.cpp [hunk 6]
```diff
@@ -748,6 +767,41 @@ void NumberFormatterApiTest::unitUsage() {
...
+            // FIXME: spaces, proper localised names
+            u"13,802 st and 7.2 lb",
+            u"1,380 st and 3.5 lb",
+            u"138 st and 0.35 lb",
+            u"13 st and 11 lb",
+            u"1 st and 5.3 lb",
+            u"1 lb and 15 oz",
+            u"0 lb and 3.1 oz",
+            u"0 lb and 0.31 oz",
+            u"0 lb and 0 oz");
```
Comment (sffc) [inline icu4c/source/test/intltest/numbertest_api.cpp:804]: Doesn't this test case make you super satisfied in all the work you've been doing? :smiley:

File: icu4c/source/test/intltest/numbertest_api.cpp [hunk 7]
```diff
@@ -677,20 +678,28 @@ void NumberFormatterApiTest::unitMeasure() {
             u"5 a\u00F1os");
 }
 
+// TODO(hugovdm): move down for consistent method order.
 void NumberFormatterApiTest::unitUsage() {
```
Comment (sffc) [inline icu4c/source/test/intltest/numbertest_api.cpp:682]: Also add tests for edge cases like a negative number as input.

File: icu4c/source/i18n/number_longnames.cpp [hunk 8]
```diff
@@ -386,6 +445,83 @@ const Modifier* LongNameHandler::getModifier(Signum /*signum*/, StandardPlural::
     return &fModifiers[plural];
```
Comment (sffc) [inline icu4c/source/i18n/number_longnames.cpp:445]: You could put a U_ASSERT in here to make sure we're not in mixed unit mode

File: icu4c/source/i18n/number_longnames.h [hunk 9]
```diff
@@ -34,58 +35,134 @@ class LongNameHandler : public MicroPropsGenerator, public ModifierStore, public
...
     void
     processQuantity(DecimalQuantity &quantity, MicroProps &micros, UErrorCode &status) const U_OVERRIDE;
 
+    // TODO(units): investigate whether we might run into Mixed Unit trouble
+    // with this. This override for ModifierStore::getModifier does not support
+    // mixed units: investigate under which circumstances it gets called.
     const Modifier* getModifier(Signum signum, StandardPlural::Form plural) const U_OVERRIDE;
 
   private:
+    // A set of pre-computed modifiers, one for each plural form.
```
Comment (sffc) [inline icu4c/source/i18n/number_longnames.h:82]: Optional: Basically what you're doing in this PR is making two versions of LongNameHandler: one for mixed units and one for single units.  fModifiers is unused in mixed unit mode, and your new fields like fMixedUnitData are unused in single unit mode.  Consider splitting these into two separate classes.

File: icu4c/source/test/intltest/numbertest_api.cpp (hunk 7 above)
Comment (hugovdm) [reply icu4c/source/test/intltest/numbertest_api.cpp:682]: Still TO-DO, tomorrow. (There are FIXMEs in this already, some cleanup also still required.)

File: icu4c/source/test/intltest/numbertest_api.cpp (hunk 6 above)
Comment (hugovdm) [reply icu4c/source/test/intltest/numbertest_api.cpp:804]: It's the first step: I'll feel much more satisfied when the tests are further diversified, to feel more convinced / to prove that it's as flexible as it should be! ;-D

Coming up...

File: icu4c/source/i18n/number_longnames.cpp [hunk 10]
```diff
@@ -200,24 +212,39 @@ UnicodeString getPerUnitFormat(const Locale& locale, const UNumberUnitWidth &wid
...
+    // `unitRef`, with `getType()` returning a zero-length string, which does
+    // actually have a "basic unit" representation.
+    U_ASSERT(fillIn != nullptr);
     if (uprv_strlen(unitRef.getType()) == 0 || uprv_strlen(perUnit.getType()) == 0) {
-        // TODO(ICU-20941): Unsanctioned unit. Not yet fully supported. Set an error code.
-        status = U_UNSUPPORTED_ERROR;
-        return;
+        // TODO(ICU-20941): also deal with compound units (not just mixed units).
+        if (unitRef.getComplexity(status) == UMEASURE_UNIT_MIXED) {
+            // TODO(review): can we simply U_ASSERT this? Passing a `perUnit`
```
[the remark this answers is not in the document]
Comment (sffc) [reply icu4c/source/i18n/number_longnames.cpp:226]: Okay, I guess it's fine to handle perUnit.  It's just an extra code path we need to think about.  My thinking was that new users should only use unit and not perUnit, so we can just limit the cases where perUnit is accepted to make the code easier to reason about.

File: icu4c/source/i18n/number_longnames.cpp [hunk 11]
```diff
@@ -295,6 +321,35 @@ void LongNameHandler::forCompoundUnit(const Locale &loc, const MeasureUnit &unit
...
+                                   const MicroPropsGenerator *parent, LongNameHandler *fillIn,
+                                   UErrorCode &status) {
+    if (fillIn == nullptr) {
+        status = U_INTERNAL_PROGRAM_ERROR;
+        return;
+    }
+
+    // TODO/FIXME: obscure "AddressSanitizer: heap-buffer-overflow" if I swap
+    // the order of these two lines:
+    LocalArray<MeasureUnit> individualUnits = unit.splitToSingleUnits(fillIn->mixedUnitCount, status);
```
[the remark this answers is not in the document]
Comment (sffc) [reply icu4c/source/i18n/number_longnames.cpp]: Hmm, the other problem here is that you would need to clone your MeasureUnitImpl in order to have something mutable on which you can call the build function.

Either leave this code as-is, or add some other function similar to findBySubType.

File: icu4c/source/i18n/number_longnames.h [hunk 12]
```diff
@@ -34,58 +35,134 @@ class LongNameHandler : public MicroPropsGenerator, public ModifierStore, public
...
+     * on the plural form.
+     *
+     * If formatting a mixed unit, `quantity` is taken as the final smallest
+     * unit, while the larger unit values must be provided via
+     * `micros.mixedMeasures`.
+     */
     void
     processQuantity(DecimalQuantity &quantity, MicroProps &micros, UErrorCode &status) const U_OVERRIDE;
 
+    // TODO(units): investigate whether we might run into Mixed Unit trouble
```
[the remark this answers is not in the document]
Comment (sffc) [reply icu4c/source/i18n/number_longnames.h:76]: I think the ImmutablePatternModifier call site of `getModifier(quantity, pluralForm)` is harmless and doesn't affect LongNamesHandler.  The main case here is NumberRangeFormatterImpl.

File: icu4c/source/i18n/number_longnames.cpp (hunk 8 above)
Comment (sffc) [reply icu4c/source/i18n/number_longnames.cpp:445]: As mentioned above, I think the main problem here is range formatting.  We should evaluate the degree to which range formatting breaks with mixed units, and how we want to fix it.

File: icu4c/source/i18n/number_longnames.h (hunk 9 above)
Comment (sffc) [reply icu4c/source/i18n/number_longnames.h:83]: I think it makes more sense and is cleaner to have the two classes.

You can put them in the same cpp file if you like, or into new cpp files.

Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.h:78]: We now have two classes, ensuring this PR doesn't change behaviour for anything other than the usage() case with mixed units (since the only entry point is currently via LongNameMultiplexer).

**Comment by hugovdm** [comment]
Ready for another review! I've split the response over 5 commits since that might make reviewing easier, depending on your reviewing style.

I also notice that when I *respond* to comment threads, it might be cleaner to just do individual comments instead of bundling them up into a "github PR review".

File: icu4c/source/i18n/number_longnames.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.h:71]: Thanks. I leave dealing with this for the TODO - particularly since this PR should now have no impact on that code path (the only entry point with this PR is via LongNameMultiplexer).

File: icu4c/source/i18n/number_longnames.cpp (hunk 8 above)
Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp:398]: (As mentioned above and below, this PR should have no impact on range formatting now.)

File: icu4c/source/i18n/number_longnames.cpp [hunk 13]
```diff
@@ -445,8 +398,49 @@ const Modifier* LongNameHandler::getModifier(Signum /*signum*/, StandardPlural::
...
+
+    // We need a localised NumberFormatter for the integers of the bigger units
+    // (providing Arabic numerals, for example).
+    fillIn->fIntegerFormatter = NumberFormatter::withLocale(loc);
+}
+
+void MixedUnitLongNameHandler::processQuantity(DecimalQuantity &quantity, MicroProps &micros,
+                                               UErrorCode &status) const {
+    U_ASSERT(fMixedUnitCount > 1);
+    if (parent != NULL) {
```
Comment (sffc) [inline icu4c/source/i18n/number_longnames.cpp:435]: nullptr

File: icu4c/source/i18n/number_longnames.cpp [hunk 14]
```diff
@@ -532,13 +535,38 @@ LongNameMultiplexer::forMeasureUnits(const Locale &loc, const MaybeStackVector<M
...
+    }
     result->fMeasureUnits.adoptInstead(new MeasureUnit[units.length()]);
     for (int32_t i = 0, length = units.length(); i < length; i++) {
         const MeasureUnit& unit = *units[i];
         result->fMeasureUnits[i] = unit;
-        // Create empty new LongNameHandler:
-        LongNameHandler *lnh = result->fLongNameHandlers.emplaceBackAndCheckErrorCode(status);
-        LongNameHandler::forMeasureUnit(loc, unit, MeasureUnit(), width, rules, NULL, lnh, status);
+        if (unit.getComplexity(status) == UMEASURE_UNIT_MIXED) {
+            // TODO(review): do we want a createAndCheckStatusCode? It would be
```
Comment (sffc) [inline icu4c/source/i18n/number_longnames.cpp:547]: That sounds like a great idea to me

**Comment by hugovdm** [comment]
Scary... I seem to have caused a failure in format/MeasureFormatTest/TestIndividualPluralFallback

         TestIndividualPluralFallback {
         FAIL: 2 deg temp in fr_CA; got ""; expected "2\u00B0" - (Are you missing data?)
         TestIndividualPluralFallback destructor: expected success but got error: U_ILLEGAL_ARGUMENT_ERROR

Investigating...

**Comment by hugovdm** [comment]
Observations - an obscure catch, via a "lucky" inefficiency:

    U_ASSERT(unitRef.getComplexity(status) != UMEASURE_UNIT_MIXED);

First observation: getComplexity is quite inefficient for built-in types - it requires converting to a MeasureUnitImpl, which means producing the identifier and then parsing it. We don't otherwise know the complexities of the built-in types...

Now the thing that I caught with this: "temperature-generic" is not parseable by MeasureUnitImpl. Never was. It also will never be in "convertUnits" because "temperature-generic" is not convertible to other temperature types.
(I checked that it was missing in the original hard-coded list too:
https://github.com/icu-units/icu/commit/84e964e00643404534087e12dfb58147bf3915fe)

File: icu4c/source/i18n/number_longnames.cpp (hunk 13 above)
Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp]: Done.

File: icu4c/source/test/intltest/measfmttest.cpp [hunk 15]
```diff
@@ -3130,11 +3130,23 @@ void MeasureFormatTest::TestIndividualPluralFallback() {
...
     UnicodeString expected = UNICODE_STRING_SIMPLE("2\\u00B0").unescape();
     UnicodeString actual;
-    assertEquals("2 deg temp in fr_CA", expected, mf.format(twoDeg.orphan(), actual, errorCode), TRUE);
+    // Formattable adopts the pointer
+    mf.format(Formattable(twoDeg.orphan()), actual, errorCode);
+    if (errorCode.errIfFailureAndReset("mf.format(...) failed.")) {
+        return;
+    }
+    assertEquals("2 deg temp in fr_CA", expected, actual, TRUE);
+    errorCode.errIfFailureAndReset("mf.format failed");
```
Comment (hugovdm) [inline icu4c/source/test/intltest/measfmttest.cpp:3149]: These changes helped me narrow down an error (I think it was the "temperature-generic doesn't parse" problem tripping a getComplexisty() assert I added).

File: icu4c/source/i18n/number_longnames.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp]: Now simpler due to splitting out MixedUnitLongNameHandler, which doesn't take a perUnit parameter.

File: icu4c/source/i18n/number_longnames.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp]: Left it as is.

File: icu4c/source/i18n/number_longnames.cpp (hunk 14 above)
Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp]: Done, now using: createAndCheckErrorCode

File: icu4c/source/test/intltest/numbertest_api.cpp [hunk 16]
```diff
@@ -118,6 +121,95 @@ void NumberFormatterApiTest::runIndexedTest(int32_t index, UBool exec, const cha
     TESTCASE_AUTO_END;
 }
 
+void NumberFormatterApiTest::microPropsInternals(void) {
```
Comment (sffc) [inline icu4c/source/test/intltest/numbertest_api.cpp:124]: I have a slight preference for these tests to go somewhere between `TESTCASE_AUTO(locale);` and the end of the test cases in this file

File: icu4c/source/test/intltest/numbertest_api.cpp [hunk 17]
```diff
@@ -678,20 +771,28 @@ void NumberFormatterApiTest::unitMeasure() {
...
+// TODO(hugovdm): move down for consistent method order.
 void NumberFormatterApiTest::unitUsage() {
-    UnlocalizedNumberFormatter unloc_formatter =
-        NumberFormatter::with().usage("road").unit(MeasureUnit::getMeter());
-
     IcuTestErrorCode status(*this, "unitUsage()");
+    UnlocalizedNumberFormatter unloc_formatter;
+    LocalizedNumberFormatter formatter;
+    FormattedNumber formattedNum;
+    StringPiece testCase;
```
Comment (sffc) [inline icu4c/source/test/intltest/numbertest_api.cpp:780]: I don't think you ever use `testCase`.  Whenever you do

```cpp
testCase = "abc";
uTestCase = testCase.data();
```

consider instead doing

```cpp
uTestData = u"abc";
```

**Review by younies — APPROVED** [review]

File: icu4c/source/i18n/number_microprops.h [hunk 18]
```diff
@@ -22,6 +22,49 @@
...
+    };
+
+    // Assignment operator
+    void operator=(const IntMeasures &rhs) {
+        int32_t length = rhs.capacity;
+        if (this->resize(length, 0) != NULL) {
+            U_ASSERT(this->capacity == rhs.capacity);
+            // FIXME: test what happens if this is swapped - have tests that would
+            // catch it?
+            uprv_memcpy(this->ptr, rhs.ptr, (size_t)length * sizeof(int64_t));
```
[the remark this answers is not in the document]
Comment (sffc) [reply icu4c/source/i18n/number_microprops.h:60]: Can you ensure there is a follow-up issue for this?

File: icu4c/source/test/intltest/numbertest_api.cpp (hunk 7 above)
Comment (sffc) [reply icu4c/source/test/intltest/numbertest_api.cpp:686]: Follow-up: #63

File: icu4c/source/test/intltest/numbertest_api.cpp [hunk 19]
```diff
@@ -917,6 +820,68 @@ void NumberFormatterApiTest::unitUsageErrorCodes() {
     status.assertSuccess();
 }
 
+void NumberFormatterApiTest::unitPipeline() {
+    IcuTestErrorCode status(*this, "unitPipeline()");
+    LocalizedNumberFormatter nf;
+    FormattedNumber num;
+
+    // Built-in unit, meter-per-second
+    nf = NumberFormatter::with().unit(MeasureUnit::getMeterPerSecond()).locale("en-GB");
```
Comment (sffc) [inline icu4c/source/test/intltest/numbertest_api.cpp:829]: Better:

```cpp
assertFormatSingle(
    u"unit/meter-per-second",
    u"measure-unit/meter-per-second",
    NumberFormatter::with().unit(MeasureUnit::getMeterPerSecond()),
    Locale("en-GB"),
    2.4,
    u"2.4 m/s"
)
```

File: icu4c/source/i18n/number_microprops.h (hunk 18 above)
[the remark this answers is not in the document]
Comment (hugovdm) [reply icu4c/source/i18n/number_microprops.h:60]: https://github.com/icu-units/icu/issues/64

File: icu4c/source/i18n/measunit.cpp [hunk 20]
```diff
@@ -604,9 +604,9 @@ static const char * const gSubTypes[] = {
 static int32_t unitPerUnitToSingleUnit[][4] = {
         {378, 382, 12, 5},
         {378, 387, 12, 6},
-        {388, 343, 19, 0},
-        {390, 350, 19, 2},
-        {392, 343, 19, 3},
+        {388, 343, 19, 0},  // kilometer per hour
+        {390, 350, 19, 2},  // meter per second
+        {392, 343, 19, 3},  // mile per hour
```
Comment (hugovdm) [inline icu4c/source/i18n/measunit.cpp:609]: I'm expecting these comments to simply be dropped when this code is next regenerated. (It will be a messy regeneration: I want to get involved with that.) I made these comments to help clarify why we're getting "meter-per-second-per-second".

File: icu4c/source/test/intltest/numbertest_api.cpp [hunk 21]
```diff
@@ -825,35 +825,46 @@ void NumberFormatterApiTest::unitPipeline() {
...
-    assertEquals("meter per second builtin", "2.4 m/s", nf.formatDouble(2.4, status).toString(status));
-    status.assertSuccess();
-
-    // Built-in unit composed of built-in per built-in
-    nf = NumberFormatter::with().unit(METER).perUnit(SECOND).locale("en-GB");
-    assertEquals("meter per second composed", "2.4 m/s", nf.formatDouble(2.4, status).toString(status));
-    status.assertSuccess();
-
-    // "forIdentifier" tries to provide a built-in unit if there is one.
+    // "forIdentifier" must provide a built-in unit if there is one, even for
```
Comment (sffc) [inline icu4c/source/test/intltest/numbertest_api.cpp:828]: This test belongs in MeasureFormatTest

Comment (hugovdm) [reply icu4c/source/test/intltest/numbertest_api.cpp]: Done: now testing a handful of interesting compound units parsed in various ways.

**Comment by hugovdm** [comment]
(Merge conflict expected, should be simple to resolve, need not block review I think.)

File: icu4c/source/i18n/number_longnames.cpp [hunk 22]
```diff
@@ -386,6 +399,125 @@ const Modifier* LongNameHandler::getModifier(Signum /*signum*/, StandardPlural::
...
+    fListFormatter->format(outputMeasuresList.getAlias(), fMixedUnitCount, premixedFormatPattern,
+                           status);
+    SimpleFormatter premixedCompiled(premixedFormatPattern, 0, 1, status);
+    if (U_FAILURE(status)) {
+        return &micros.helpers.emptyWeakModifier;
+    }
+
+    // Return a SimpleModifier for the "premixed" pattern
+    micros.helpers.mixedUnitModifier =
+        SimpleModifier(premixedCompiled, {UFIELD_CATEGORY_NUMBER, UNUM_MEASURE_UNIT_FIELD}, false,
```
Comment (sffc) [inline icu4c/source/i18n/number_longnames.cpp:507]: Please file an issue to remind us to think harder about what field positions we want.

Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp:507]: https://github.com/icu-units/icu/issues/67

Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp:507]: Thanks, merging!
